[ML] AIOps Log Rate Analysis: improves explanation of log rate spike/dip

alvarezmelissa87 opened the pull request:
## Summary

Related issue: https://github.com/elastic/kibana/issues/182714

This PR adds a `Log rate change` column to Log rate analysis results table. 

The log rate change is calculated by getting the number of buckets for baseline/deviation using the timerange and interval and then comparing the average rates per bucket for baseline vs deviation.

<img width="1466" alt="image" src="https://github.com/elastic/kibana/assets/6446462/9b2f1b80-d1d5-407e-908c-f611e54be4f3">

<img width="1471" alt="image" src="https://github.com/elastic/kibana/assets/6446462/56cb2c35-3758-4b24-9f50-2f99242af1b3">



### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

walterra commented on x-pack/plugins/aiops/public/components/log_rate_analysis_results_table/get_baseline_and_deviation_rates.ts:
```diff
@@ -0,0 +1,58 @@
+  const logRateChange =
+    baselineBucketRate > 0
+      ? analysisType === LOG_RATE_ANALYSIS_TYPE.SPIKE
+        ? `${Math.round((deviationBucketRate / baselineBucketRate) * 100) / 100}x higher`
+        : `${Math.round((baselineBucketRate / deviationBucketRate) * 100) / 100}x lower`
```
Guess we also need to add i18n to those strings.

peteharverson replied:
Can you use https://github.com/elastic/kibana/blob/main/x-pack/plugins/ml/common/util/metric_change_description.ts for these, as used for the anomalies table? That will use 'More than 100x higher' if the factor is `> 100`, which is probably appropriate to use here too? Or if you can't call that same code, then use the same rules around precision so that e.g. `x14.23 higher` becomes `x14 higher`.

alvarezmelissa87 replied:
Updated to round and to add localization in ff861eb3f280d3812b7536fcbabcd06eced039cc

walterra commented on x-pack/plugins/aiops/public/components/log_rate_analysis_results_table/get_baseline_and_deviation_rates.ts:
```diff
@@ -0,0 +1,58 @@
+) {
+  const logRateChange =
+    baselineBucketRate > 0
+      ? analysisType === LOG_RATE_ANALYSIS_TYPE.SPIKE
```
Our style guide recommends not doing nested ternaries. I'd suggest just using if/else in this function and return early if something matches.

alvarezmelissa87 replied:
Updated in ff861eb3f280d3812b7536fcbabcd06eced039cc

walterra commented on x-pack/plugins/aiops/public/components/log_rate_analysis_results_table/log_rate_analysis_results_table.tsx:
```diff
@@ -83,6 +83,11 @@ export const LogRateAnalysisResultsTable: FC<LogRateAnalysisResultsTableProps> =
   const selectedSignificantItem = useAppSelector(
     (s) => s.logRateAnalysisTableRow.selectedSignificantItem
   );
+  const {
+    analysisType,
+    windowParameters,
+    documentStats: { documentCountStats },
+  } = useAppSelector((s) => s.logRateAnalysis);
```
Since `useColumns` is a hook, you could move this inside the hook itself so doesn't need to get passed on as arguments to `useColumns`.

walterra commented on x-pack/plugins/aiops/public/components/log_rate_analysis_results_table/log_rate_analysis_results_table_groups.tsx:
```diff
@@ -93,6 +93,11 @@ export const LogRateAnalysisResultsGroupsTable: FC<LogRateAnalysisResultsTablePr
 
   const pinnedGroup = useAppSelector((s) => s.logRateAnalysisTableRow.pinnedGroup);
   const selectedGroup = useAppSelector((s) => s.logRateAnalysisTableRow.selectedGroup);
+  const {
+    analysisType,
+    windowParameters,
+    documentStats: { documentCountStats },
+  } = useAppSelector((s) => s.logRateAnalysis);
```
Since `useColumns` is a hook, you could move this inside the hook itself so doesn't need to get passed on as arguments to `useColumns`.

alvarezmelissa87 replied:
Good catch - moved in ff861eb3f280d3812b7536fcbabcd06eced039cc

peteharverson commented on x-pack/plugins/aiops/public/components/log_rate_analysis_results_table/use_columns.tsx:
```diff
@@ -273,9 +332,148 @@ export const useColumns = (
         sortable: true,
         valign: 'middle',
       },
+      ['Baseline rate']: {
```
Can the `Baseline rate` column be made sortable?

alvarezmelissa87 replied:
Updated in ff861eb3f280d3812b7536fcbabcd06eced039cc

peteharverson commented on x-pack/plugins/aiops/public/components/log_rate_analysis_results_table/use_columns.tsx:
```diff
@@ -273,9 +332,148 @@ export const useColumns = (
+        sortable: false,
+        valign: 'middle',
+      },
+      ['Deviation rate']: {
```
Can the `Deviation rate` column be made sortable?

alvarezmelissa87 replied:
Updated in ff861eb3f280d3812b7536fcbabcd06eced039cc

peteharverson commented on x-pack/plugins/aiops/public/components/log_rate_analysis_results_table/use_columns.tsx:
```diff
@@ -273,9 +332,148 @@ export const useColumns = (
+        sortable: false,
+        valign: 'middle',
+      },
+      ['Log rate change']: {
```
Can the `Log rate change` column be made sortable? I think the similar column in the anomalies table sorts by size of factor of increase / decrease.

alvarezmelissa87 replied:
Updated in ff861eb3f280d3812b7536fcbabcd06eced039cc
Since we have the pValue - which is an indication of the impact - I used that to sort 👍

peteharverson replied:
Dosen't look like `pValue` always correlates to the size of the change.

<img width="727" alt="Screenshot 2024-06-21 at 11 54 38" src="https://github.com/elastic/kibana/assets/7405507/d6f47d68-0f60-4a42-9abf-93e241727c19">


I think the `sortable` prop can be `((item: T) => Primitive)` as well as a simple boolean and then you can use logic like https://github.com/elastic/kibana/blob/main/x-pack/plugins/ml/server/models/results_service/build_anomaly_table_items.js#L105 to return the factor difference (or maybe you already know the change up front?).

peteharverson commented on x-pack/plugins/aiops/public/components/log_rate_analysis/log_rate_analysis_results.tsx:
```diff
@@ -186,7 +186,11 @@ export const LogRateAnalysisResults: FC<LogRateAnalysisResultsProps> = ({
-  const [skippedColumns, setSkippedColumns] = useState<ColumnNames[]>(['p-value']);
+  const [skippedColumns, setSkippedColumns] = useState<ColumnNames[]>([
+    'p-value',
+    'Baseline rate',
```
@alvarezmelissa87 @walterra what do you think about showing the Baseline and Deviation rate columns by default, and hiding the Doc count? I like to see these columns in combination with the 'Log rate change' column.

alvarezmelissa87 replied:
🤔 I think that could work but we'd need all 3 showing up since the groups table and expanded row hare the same column controls. Might need to think about it a bit more if we want to make them independent.

peteharverson replied:
I'm happy to leave it as it is, with baseline and deviation rate hidden by default. The more I think about it, the less I like the idea of adding two extra numeric columns into the table in the default view.

peteharverson commented on x-pack/plugins/aiops/public/components/log_rate_analysis_results_table/get_baseline_and_deviation_rates.ts:
```diff
@@ -0,0 +1,58 @@
+      : i18n.translate(
+          'xpack.aiops.logRateAnalysis.resultsTableGroups.logRateDecreaseLabelColumnTooltip',
+          {
+            defaultMessage: '0 docs from ${baselineBucketRate} in baseline',
```
Need to remove the `$` here, but looks like there is something wrong with the logic as `baselineBucketRate` for a dip is showing as `0` in my tests:

<img width="240" alt="Screenshot 2024-06-19 at 17 00 21" src="https://github.com/elastic/kibana/assets/7405507/d7cbb8c3-132c-49e2-b6eb-258591e780f6">

alvarezmelissa87 replied:
Fixed the left in $ from the string interpolation in ff861eb3f280d3812b7536fcbabcd06eced039cc
Looking into the logic now 👍 Good catch!

alvarezmelissa87 replied:
Fixed the calculation in 2686407c0070e1c6d9024c2df9462397f91f4253

peteharverson commented on x-pack/plugins/aiops/public/components/log_rate_analysis/log_rate_analysis_results.tsx:
```diff
@@ -186,7 +186,11 @@ export const LogRateAnalysisResults: FC<LogRateAnalysisResultsProps> = ({
   const [shouldStart, setShouldStart] = useState(false);
   const [toggleIdSelected, setToggleIdSelected] = useState(resultsGroupedOffId);
-  const [skippedColumns, setSkippedColumns] = useState<ColumnNames[]>(['p-value']);
+  const [skippedColumns, setSkippedColumns] = useState<ColumnNames[]>([
```
Not related to the changes in this PR, but I think it would be a nice enhancement to persist the user's choice of selected columns to e.g. local storage. I found myself adding in Baseline and Deviation rates for display every time I reran an analysis.

peteharverson replied:
Added as an item to https://github.com/elastic/kibana/issues/181111.

alvarezmelissa87 commented:
@elasticmachine merge upstream

alvarezmelissa87 commented:
This has been updated and is ready for another look when you get a chance 🙏  cc @peteharverson, @walterra

peteharverson reviewed (commented):
Gave the latest changes a test. All looks good - only item to look at is the way sorting on the change column works.

peteharverson commented on x-pack/plugins/aiops/public/components/log_rate_analysis_results_table/get_baseline_and_deviation_rates.ts:
```diff
@@ -0,0 +1,96 @@
+      message = i18n.translate(
+        'xpack.aiops.logRateAnalysis.resultsTableGroups.logRateDocIncreaseLabel',
+        {
+          defaultMessage: '{deviationBucketRate} docs rate up from 0 in baseline',
```
`docs` needs to handle the singular case as it can be `1`.

walterra reviewed (approved):
Latest changes LGTM.

alvarezmelissa87 commented:
> Gave the latest changes a test. All looks good - only item to look at is the way sorting on the change column works.

Thanks for taking a look, @peteharverson 🙏  Adding sorting for that computed Log rate change is a can of worms as it requires us to make some changes to the custom sorting we've got going on now and will likely require a bit of code churn.
For now, I've removed sorting for that column and added it as a follow up in https://github.com/elastic/kibana/issues/181111

peteharverson reviewed (approved):
Latest changes LGTM
